### ISSUE TITLE
Ganesha exports

### ISSUE THREAD

**Issue by SergeyLysanov** [nfs-ganesha/nfs-ganesha#105]
Hello!

I'm using ganesha v.2.3.0 and have some issues. I try to add a ploop export dynamically.

So, we have mounted ploop:
`
[root@localhost nfs]# mount
/dev/ploop29844p1 on /vstorage-nfs/svc_name/export_name type ext4 (rw,relatime,data=ordered,balloon_ino=12)
`

Start ganesha and try to add export:

```
[root@localhost nfs]# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/etc/ganesha/svc_name/export_name.conf string:"export(export_id=12485)"
Error org.freedesktop.DBus.Error.InvalidFileContent: 0 export entries in /etc/ganesha/svc_name/export_name.conf added because (export create, block validation) errors. Details:
Config File (/etc/ganesha/svc_name/export_name.conf:12): 1 validation errors in block FSAL
Config File (/etc/ganesha/svc_name/export_name.conf:12): Errors processing block (FSAL)
Config File (/etc/ganesha/svc_name/export_name.conf:1): 1 validation errors in block EXPORT
Config File (/etc/ganesha/svc_name/export_name.conf:1): Errors found in configuration block EXPORT

[root@localhost nfs]# less /var/log/ganesha.log
28/06/2016 18:31:27 : epoch 577297d0 : localhost.localdomain : ganesha.nfsd-1473[dbus_heartbeat] claim_posix_filesystems :FSAL :CRIT :No file system for export path /vstorage-nfs/svc_name/export_name
28/06/2016 18:31:27 : epoch 577297d0 : localhost.localdomain : ganesha.nfsd-1473[dbus_heartbeat] vfs_create_export :FSAL :CRIT :claim_posix_filesystems(/vstorage-nfs/svc_name/export_name) returned No such file or directory (2)
28/06/2016 18:31:27 : epoch 577297d0 : localhost.localdomain : ganesha.nfsd-1473[dbus_heartbeat] fsal_commit :CONFIG :CRIT :Could not create export for (/vstorage-nfs/svc_name/export_name) to (/vstorage-nfs/svc_name/export_name)
28/06/2016 18:31:27 : epoch 577297d0 : localhost.localdomain : ganesha.nfsd-1473[dbus_heartbeat] gsh_export_addexport :EXPORT :CRIT :0 export entries in /etc/ganesha/svc_name/export_name.conf added because (export create, block validation) errors
```

As you can see, ganesha can't determine the type of the filesystem. We should restart ganesha after ploop mount for adding export.

Second issue occurs when I remove export and try to unmount ploop. I get the following message:

```
[root@localhost nfs]# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.RemoveExport uint16:1

[root@localhost nfs]# ploop umount /vstorage/my_pcs4/services/svc_name/shares/export_name/DiskDescriptor.xml
Unmounting file system at /vstorage-nfs/svc_name/export_name
Unmounting device /dev/ploop29844
--- lsof /dev/ploop29844 output ---
COMMAND    PID USER   FD   TYPE     DEVICE SIZE/OFF  NODE NAME
ganesha.n 4337 root    6r   BLK 182,477504  0t17408 28915 /dev/ploop29844
ploop     4540 root    5r   BLK 182,477504      0t0 28915 /dev/ploop29844
sh        4558 root    5r   BLK 182,477504      0t0 28915 /dev/ploop29844
--- lsof finished ---
Error in ploop_stop (ploop.c:995): PLOOP_IOC_STOP: Device or resource busy
Opening delta /vstorage/my_pcs4/services/svc_name/shares/export_name/image
Error in open_delta (delta_read.c:167): Image is in use /vstorage/my_pcs4/services/svc_name/shares/export_name/image
Error in ploop_umount (ploop.c:2310): Warning: saving format extension failed: 4: Device or resource busy
```

After removing export ganesha keep file descriptor of the /dev/ploop and continues to use mount.

To solve first issue I offer to parse filesystem and search new entry in /etc/mtab on each vfs_create_export call. Now it proceed only on start of ganesha.  

I have no idea about second issue yet. Why it happened? 

Thanks!

**Comment by ffilz** [comment]
I don't think we want to parse /etc/mtab each time, and reparsing would cause some issues with already claimed filesystems. Also Ganesha is not well prepared to handle the addition of a sub-mounted export, for example if you had an export:

/export/fs1

And then wanted to dynamically add an export;

/export/fs1/some/path/fs2

It PROBABLY will work ok if fs2 did not exist originally. It definitely won't if fs2 already exists and is actively cached (and especially if it has active state open/locks/delegations) via the first export.

Where we probably should re-parse mtab is on dBus add export command. But we still need to take some effort to make sure it doesn't upset existing exports.

As to why unexport didn't allow you to unmount, there is probably a refcount leak such that we never close the fd on the export root.

**Comment by SergeyLysanov** [comment]
I found a refcount leak of gsh_export struct.

The leak of refcount occurs in second AddExport call. So, after that RemoveExport command doesn't free export in put_gsh_export function:
```
root@slysanov-pc:/home/slysanov# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/etc/ganesha/exports/export.conf string:"export(export_id=33)"
method return sender=:1.129 -> dest=:1.130 reply_serial=2
   string "1 exports added"
root@slysanov-pc:/home/slysanov# 
root@slysanov-pc:/home/slysanov# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/etc/ganesha/exports/export.conf string:"export(export_id=33)"
Error org.freedesktop.DBus.Error.InvalidFileContent: Selected entries in /etc/ganesha/exports/export.conf already active!!!

root@slysanov-pc:/home/slysanov# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.RemoveExport uint16:33
method return sender=:1.129 -> dest=:1.132 reply_serial=2

```
This is a piece of code where refcount is incremented, but not decremented:
```
static int export_commit_common()  {
 ...
 probe_exp = get_gsh_export_by_path(export->fullpath, true);
 if (probe_exp != NULL &&
    export->pseudopath == NULL &&
    export->FS_tag == NULL) {
	LogCrit(COMPONENT_CONFIG,
		"Duplicate path (%s) without unique tag or Pseudo path",
		export->fullpath);
	err_type->invalid = true;
	errcnt++;
	put_gsh_export(probe_exp);
 }
 ...
}
```

I don't understand the meaning of  export->pseudopath == NULL condition.  IMHO if we have duplicate we should put refcount in any case.

**Comment by ffilz** [comment]
Yea, we should always put_gsh_export(probe_exp); if it is non-NULL...

Good catch. I will fix it and get it in the next 2.4 build (and then it can be backported to 2.3 if necessary)

You can also avoid the problem for now by not exporting the same Path with different Tag or Pseudo...

**Comment by SergeyLysanov** [comment]
If you don't mind I will fix it and send you patch. So, we should use only probe_exp != NULL condition.
I also will send a patch with re-parse mtab on AddExport command  for review.

Thanks!

**Comment by ffilz** [comment]
Already fixed and pushed to gerrit:

https://review.gerrithub.io/282486

**Comment by SergeyLysanov** [comment]
I don't see an elegant solution to re-parse /etc/mtab only on AddExport command and don't do it on startup.

What do you think about new dBus command for re-scanning mounted file systems? Something like PopulateFS?

**Comment by ffilz** [comment]
We could definitely have a dBus command to force a rescan, but the mechanism to do that could also/instead be called by dBus add export...

**Comment by SergeyLysanov** [comment]
@wyllys66

addExport issue has been fixed already and merged to the next branch: 
 https://gerrithub.io/#/c/282540/

**Comment by wyllys66** [comment]
Thank you!

**Comment by wyllys66** [comment]
Will the fix make it into the next 2.3.3 release?  If not, can it?
